File: python_mid_june_2024/20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = deque()

        for c in s:
            if c == ')' or c ==']' or c =='}':
                if not stack:
                    return False
                cpop = stack.pop()
                if c == ')' and cpop != '(':
                    return False
                if c == ']' and cpop != '[':
                    return False
                if c == '}' and cpop != '{':
                    return False
            else:
                stack.append(c)
        
        return not stack

File: java_july/20. Valid Parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for (char c : s.toCharArray()) {
            if (c == ')' || c == ']' || c == '}') {
                if (stack.isEmpty()) {
                    return false;
                }
                char pop = stack.pop();
                if ((c == ')' && pop != '(') || (c == ']' && pop != '[') || (c == '}' && pop != '{')) {
                    return false;
                }
            }
            else {
                stack.push(c);
            }
        }
        return stack.isEmpty();
    }
}